// sapsar/src/util/SapsarErrorPage.js
const Combine = require("../basic/Combine")
const { head, title, style, body, div, h1, h4, p } = require("../base/index")
const SapsarLogo = require("./SapsarLogo")


function SapsarErrorPage(msg1, msg2, msg3, msg4){
    //remove everything from error stack after "at SapsarCompiler"

    return (
        Combine(
            head(
                title(
                    "Sapsar"
                ),
                style(
                    `
                        * {
                            padding: 0;
                            margin: 0;

                        }

                        body {
                            font-family: sans-serif;
                            display: grid;
                            min-height: 100vh;
                            place-items: center;

                        }

                        .flex {
                            display: flex;
                            flex-direction: column;
                        }

                        .errorBox {
                            padding: 2rem;
                            border-radius: 2rem;
                            box-shadow: 0 0 3rem rgba(0,0,0,0.2);
                            gap: .5rem;
                            max-width: 50rem;
                            background: white;
                            position: relative;
                        }

                        .errorBox::after {
                            content: "";
                            position: absolute;
                            z-index: -1;
                            top: -.25rem;
                            left: 0;
                            right: 0;
                            bottom: -.25rem;
                            border-radius: 2rem;
                            background: linear-gradient(90deg, red, orange);
                        }

                        svg {
                            width: 22rem;
                        }


                        .message {
                            background: #000;
                            border-radius: 10rem;
                            width: max-content;
                            padding: 1rem 2rem;
                            color: #fff;
                        }
                    `
                )
            ),
            body(
                div(
                    SapsarLogo,
                    div(
                        div(
                            h4(
                                msg1
                            ),
                            {
                                class: "message"
                            }
                        ),
                        h1(
                            `${msg2} - ${msg3}`
                        ),
                        p(
                            msg4
                        ),
                        {
                            class: "errorBox flex"
                        }
                    ),
                    {
                        class: "flex",
                        style: "gap: 1rem; align-items: center;"
                    }
                )
                
            )
        )
    )
}

module.exports = SapsarErrorPage
