// sapsar/src/util/SapsarCompiler.js
const { doctype, head, html, body, meta, style } = require("sapsar/base")

const Log = require("./Log")


const cacheFormat = {
    css: {
        global: ''
    },
    js: {

    },


    layout: null,
    pageCompilers: {

    },
    head: {
        '*': ' '
    },
    activeHead: {
        
    },
    reports: {
        css: {
            '*': []
        }
    },
    pages: {

    },
    staticPageRequests: []
}




function handleHead(page){
    let content = ""

    if (cache.head[page] != undefined){
        content += cache.head[page]
    }

    if (cache.activeHead[page] != undefined){
        content += cache.activeHead[page]
    }

    if (cache.head['*'] != undefined){
        content += cache.head['*']
    }


    return content
}



let cache = JSON.parse(JSON.stringify(cacheFormat))



function handleCSS(page){
    let struct = ''

    for (let x = 0; x < cache.reports.css[page].length; x++){
        let component = cache.reports.css[page][x]

        if (cache.css[component] != undefined){
            struct += style(
                cache.css[component],
                {
                    "data-sapsar-component": component
                }
            )
        }
    }

    for (let x = 0; x < cache.reports.css['*'].length; x++){
        let component = cache.reports.css['*'][x]

        if (cache.css[component] != undefined){
            struct += style(
                cache.css[component],
                {
                    "data-sapsar-component": component
                }
            )
        }
    }


    return struct
}



 function clearCacheData(){
    cache = JSON.parse(JSON.stringify(cacheFormat))
}



function renderPageStruct(page, content){
    return `
        ${doctype()}
            ${html(

                head(
                    handleHead(page),
                    meta({name:"viewport",content:"width=device-width, initial-scale=1.0"}),
                    meta({charset:"UTF-8"}),
                    handleCSS(page)
                ),
                body(
                    content
                ),

                
                {
                    lang: "en"
                },
            )}
    `
}





// Compiler Code
async function SapsarCompiler(page, data){
    if (!cache.pages[page]) {

        let staticPage = false;

        // import layout
        if (!cache.layout) {
            Log.compiler(`Generating LAYOUT CACHE at first request (${page})...`)
            cache.layout = (await import(`../../../../pages/_layout.js`)).default
        }

        // import page
        if (!cache.pageCompilers[page]) {
            Log.compiler(`Generating PAGE FUNCTION CACHE for ${page}...`)
            cache.pageCompilers[page] = (await import(`../../../../pages/${page}.js`)).default
        }

        if (cache.staticPageRequests.includes(page)) {
            staticPage = true;
        }

        
        //import head cache

        if (!cache.head[page]) {
            Log.compiler(`Generating HEAD CACHE for ${page}...`)
            cache.head[page] = ' '
        }

        //import active head cache

        if (!cache.activeHead[page]) {
            Log.compiler(`Generating ACTIVE HEAD CACHE for ${page}...`)
            cache.activeHead[page] = ' '
        }
        
        // generate css report
        if (!cache.reports.css[page]) {
            Log.compiler(`Generating CSS REPORT for ${page}...`)
            cache.reports.css[page] = []
        }







        let struct = null;

        if (staticPage){
            Log.compiler(`Generating STATIC PAGE CACHE for ${page}...`)
            const Rendered_Page = await cache.pageCompilers[page](data)
            struct = await renderPageStruct(page, Rendered_Page)

            cache.pages[page] = struct

            return struct
        }
        else {
            const Rendered_Page = await cache.pageCompilers[page](data)
            const struct = await renderPageStruct(page, Rendered_Page)


            //delete active data
            cache.activeHead[page] = ' '

            return struct
        }





    }

    else {
        // returned cached page
        return cache.pages[page]
    }
}

// export compiler (as default) and cache

module.exports = {
    SapsarCompiler,
    cache,
    clearCacheData
}
